Add recipe for portage-modes

OpenSauce04 opened the pull request:
### Brief summary of what the package does

Adds several major modes which provide syntax highlighting when viewing or editing config files for Portage, the Gentoo Linux package manager

### Direct link to the package repository

https://github.com/OpenSauce04/portage-modes

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

None needed

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
  - Warns about a lack of a `lexical-binding` directive but is otherwise fine
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

This is my first time writing anything in Elisp (or any other Lisp), so if I have done anything incorrectly or have taken a particularly strange approach to anything in the code of the package, assume it was simply due to my lack of knowledge and tell me what I can do to improve it.

riscy commented:
Thanks for this.  Largely this looks good, but here's a quick first pass -

- `;;; Commentary` is much wider than 80 characters (try `M-q` on this)
- `defface portage-modes-green-face` (also `-red`): You should make these robust to situations where the user has a green background, usually by preferring to inherit from another face instead.  Also calling a face the `green-face` will seem strange when someone customizes it to be e.g. purple.  Better to use semantic naming e.g. "warning-face" or "correct-face" for example.
- You should strive to be precise with your `auto-mode-alist` regexps: terminate with a `\\'` (end-of-string rather than end-of-line) when possible and consider what could happen if e.g. someone created a file called `^/etc/portage/package.license.diff` (or similar)

OpenSauce04 commented:
> You should make these robust to situations where the user has a green background, **usually by preferring to inherit from another face instead.**

Just to clarify, this a requirement or a suggestion? I'd like to maintain the absolute red and green of the colours if possible because the instances where they are used refer to additions and removals. It's an intuitive colour scheme that is used commonly for things such as this.

If not, what is a face that you would suggest pulling from instead?

riscy commented:
> It's an intuitive colour scheme that is used commonly for things such as this.

I understand but what you might be surprised how weird choices like this can look in the context of other color schemes (e.g. monochrome, low contrast, color-blind-friendly, or just slightly different palettes).

Naming the variables `-added` instead of `-green` and `-removed` instead of `-red` goes halfway.

Inheriting from existing faces means that your choices will "fit" with existing palettes for free.  For example my `diff-removed` face isn't red, it's `#ff665c` as dictated by the current theme I'm using.

> what is a face that you would suggest pulling from instead?

`diff-added` and `diff-removed` are solid because they are builtin.  You may need to `(require 'diff-mode)` but I think this is an appropriate dependency, even if just for the colors, since you are literally highlighting diffs.

riscy commented:
Incidentally - I don't want to come off as harsh or defensive here, but I think you should think about how you said:

> assume it was simply due to my lack of knowledge and tell me what I can do to improve it.

and then started to resist (imo) good advice:

> It's an intuitive colour scheme that is used commonly for things such as this.  If it's non-negotiable,

OpenSauce04 commented:
> Incidentally - I don't want to come off as harsh or defensive here, but I think you should think about how you said:
> 
> > assume it was simply due to my lack of knowledge and tell me what I can do to improve it.
> 
> and then started to resist (imo) good advice:
> 
> > It's an intuitive colour scheme that is used commonly for things such as this.  If it's non-negotiable,

Oh I wasn't trying to resist what you said, I was just explaining the thought process behind why I made the decision I made initially. Sorry if it came off that way.

> `diff-added` and `diff-removed` are solid because they are builtin

Oh! I didn't know these faces existed; these make way more sense. I looked over some existing options before going to custom faces but somehow missed these, thanks for bringing them up.

OpenSauce04 commented:
After I got back to my computer and actually looked at the colours of these diff-mode faces, I noticed they are significantly lighter than I expected them to be. In fact when the colour is used for the foreground it is almost invisible in the default Emacs theme:
Default:
<img src=https://github.com/user-attachments/assets/5b4f5f58-0ba7-4175-a43f-287a0f6c42e7 width=800px/>
My theme:
<img src=https://github.com/user-attachments/assets/75bd30ed-5fb4-4165-a456-9706af15af3b width=800px/>
I took a look through `M-x list-faces-display` and discovered the `error` and `success` faces, which I have opted use instead:
Default:
<img src=https://github.com/user-attachments/assets/4813d622-f337-4121-9e61-83a4d1a1856e width=800px/>
My theme:
<img src=https://github.com/user-attachments/assets/e435a603-994b-4d49-9bc1-bcd875f48c52 width=800px/>

Assuming that there are no issues with the above, I believe I have addressed the issues you mentioned originally. Please check over my changes and let me know if there is anything else you would like me to adjust. 🙂

riscy commented:
I should have looked into it, but it looks like e.g. `diff-removed` only specifies a background color in most cases (depending on the terminal):
```elisp
(defface diff-removed
  '((default
     :inherit diff-changed)
    (((class color) (min-colors 257) (background light))
     :background "#ffeeee" :extend t)
    (((class color) (min-colors 88) (background light))
     :background "#ffdddd" :extend t)
    (((class color) (min-colors 88) (background dark))
     :background "#553333" :extend t)
    (((class color))
     :foreground "red" :extend t))
  "`diff-mode' face used to highlight removed lines.")
```
and in your screenshots it looks like that background color was used as a foreground color somehow.

> I believe I have addressed the issues you mentioned originally.

Yes, looks good!
